### ISSUE TITLE
Kafka Benchmark Tests Erroring out

### ISSUE THREAD

**Issue by terrysiu-pm** [redpanda-data/openmessaging-benchmark#19]
Hi,

I'm getting the same errors as described in this [closed issue](https://github.com/redpanda-data/openmessaging-benchmark/issues/16). Per the closed issue, I updated the `org.hdrhistogram.HdrHistogram` dependency to `2.1.12`.  I also had to comment out the `com.mycila.license-maven-plugin` in the parent pom as the build would complain of various files missing license agreement wording. I ran `mvn install` and everything built successfully. I then launched two workers locally (on different ports) and then the driver locally using Kafka and the workload, `1-topic-1-partition-100b.yaml` which is set to run for 15 minutes. After about 2 minutes, the producer starts dumping out similar WARN level messages as the linked issue:

```text
09:03:14.712 [kafka-producer-network-thread | producer-1] WARN - Write error on message
java.util.concurrent.CompletionException: java.lang.ArrayIndexOutOfBoundsException: value 67283144 outside of histogram covered range. Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 1311060 out of bounds for length 1310720
        at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314) ~[?:?]
        at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319) ~[?:?]
        at java.util.concurrent.CompletableFuture$UniRun.tryFire(CompletableFuture.java:787) ~[?:?]
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506) [?:?]
        at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2073) [?:?]
        at io.openmessaging.benchmark.driver.kafka.KafkaBenchmarkProducer.lambda$sendAsync$0(KafkaBenchmarkProducer.java:49) [driver-kafka-0.0.1-SNAPSHOT.jar:?]
        at org.apache.kafka.clients.producer.KafkaProducer$InterceptorCallback.onCompletion(KafkaProducer.java:1363) [kafka-clients-2.6.0.jar:?]
        at org.apache.kafka.clients.producer.internals.ProducerBatch.completeFutureAndFireCallbacks(ProducerBatch.java:228) [kafka-clients-2.6.0.jar:?]
        at org.apache.kafka.clients.producer.internals.ProducerBatch.done(ProducerBatch.java:197) [kafka-clients-2.6.0.jar:?]
        at org.apache.kafka.clients.producer.internals.Sender.completeBatch(Sender.java:653) [kafka-clients-2.6.0.jar:?]
        at org.apache.kafka.clients.producer.internals.Sender.completeBatch(Sender.java:634) [kafka-clients-2.6.0.jar:?]
        at org.apache.kafka.clients.producer.internals.Sender.handleProduceResponse(Sender.java:554) [kafka-clients-2.6.0.jar:?]
        at org.apache.kafka.clients.producer.internals.Sender.lambda$sendProduceRequest$0(Sender.java:743) [kafka-clients-2.6.0.jar:?]
        at org.apache.kafka.clients.ClientResponse.onComplete(ClientResponse.java:109) [kafka-clients-2.6.0.jar:?]
        at org.apache.kafka.clients.NetworkClient.completeResponses(NetworkClient.java:566) [kafka-clients-2.6.0.jar:?]
        at org.apache.kafka.clients.NetworkClient.poll(NetworkClient.java:558) [kafka-clients-2.6.0.jar:?]
        at org.apache.kafka.clients.producer.internals.Sender.runOnce(Sender.java:325) [kafka-clients-2.6.0.jar:?]
        at org.apache.kafka.clients.producer.internals.Sender.run(Sender.java:240) [kafka-clients-2.6.0.jar:?]
        at java.lang.Thread.run(Thread.java:829) [?:?]
Caused by: java.lang.ArrayIndexOutOfBoundsException: value 67283144 outside of histogram covered range. Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 1311060 out of bounds for length 1310720
        at org.HdrHistogram.AbstractHistogram.handleRecordException(AbstractHistogram.java:571) ~[HdrHistogram-2.1.12.jar:2.1.12]
        at org.HdrHistogram.AbstractHistogram.recordSingleValue(AbstractHistogram.java:563) ~[HdrHistogram-2.1.12.jar:2.1.12]
        at org.HdrHistogram.AbstractHistogram.recordValue(AbstractHistogram.java:467) ~[HdrHistogram-2.1.12.jar:2.1.12]
        at org.HdrHistogram.Recorder.recordValue(Recorder.java:136) ~[HdrHistogram-2.1.12.jar:2.1.12]
        at io.openmessaging.benchmark.worker.LocalWorker.lambda$submitProducersToExecutor$8(LocalWorker.java:266) ~[classes/:?]
        at java.util.concurrent.CompletableFuture$UniRun.tryFire(CompletableFuture.java:783) ~[?:?]
```

I am on JDK11. Due to this error, the `Pub rate` starts dropping until it hits zero. Also noticed these errors as well in the producer:

```text
09:04:48.734 [kafka-producer-network-thread | producer-1] WARN - Write error on message
java.util.concurrent.CompletionException: org.apache.kafka.common.errors.TimeoutException: Expiring 1180 record(s) for test-topic-eYYvPlM-0000-0:123237 ms has passed since batch creation
        at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:331) ~[?:?]
        at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:346) ~[?:?]
        at java.util.concurrent.CompletableFuture$UniRun.tryFire(CompletableFuture.java:777) ~[?:?]
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506) [?:?]
        at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2088) [?:?]
        at io.openmessaging.benchmark.driver.kafka.KafkaBenchmarkProducer.lambda$sendAsync$0(KafkaBenchmarkProducer.java:47) [driver-kafka-0.0.1-SNAPSHOT.jar:?]
        at org.apache.kafka.clients.producer.KafkaProducer$InterceptorCallback.onCompletion(KafkaProducer.java:1363) [kafka-clients-2.6.0.jar:?]
        at org.apache.kafka.clients.producer.internals.ProducerBatch.completeFutureAndFireCallbacks(ProducerBatch.java:231) [kafka-clients-2.6.0.jar:?]
        at org.apache.kafka.clients.producer.internals.ProducerBatch.done(ProducerBatch.java:197) [kafka-clients-2.6.0.jar:?]
        at org.apache.kafka.clients.producer.internals.Sender.failBatch(Sender.java:676) [kafka-clients-2.6.0.jar:?]
        at org.apache.kafka.clients.producer.internals.Sender.sendProducerData(Sender.java:381) [kafka-clients-2.6.0.jar:?]
        at org.apache.kafka.clients.producer.internals.Sender.runOnce(Sender.java:324) [kafka-clients-2.6.0.jar:?]
        at org.apache.kafka.clients.producer.internals.Sender.run(Sender.java:240) [kafka-clients-2.6.0.jar:?]
        at java.lang.Thread.run(Thread.java:829) [?:?]
Caused by: org.apache.kafka.common.errors.TimeoutException: Expiring 1180 record(s) for test-topic-eYYvPlM-0000-0:123237 ms has passed since batch
```

Any guidance on how I might be able to get this sample workload to pass?

Thanks!

**Comment by tmgstevens** [comment]
We suspect that this is caused by https://github.com/HdrHistogram/HdrHistogram/pull/180, which doesn't appear to be upstream.

I've created a shaded version with the fix that you can try out. It's on a branch at the moment with a whole host of other fixes as well, but you're welcome to try it out and let me know how you get on.

See https://github.com/tmgstevens/openmessaging-benchmark/tree/kafka-3.2
The commit is this one: https://github.com/tmgstevens/openmessaging-benchmark/commit/ece96e8f5e73b34ec2c01465db9467415921c694 (which you could just test in isolation as well)

**Comment by terrysiu-pm** [comment]
Hi @tmgstevens , I just tried out your fork and branch (did a clean and install), ran the same workload, and the same `ArrayIndexOutOfBoundsException` occurs, eventually, replaced with the `Expired n record(s)...` error.

**Comment by tmgstevens** [comment]
@terrysiu-pm we believe this is fixed now by #35
